include/ccf/ds/attestation_sev_snp.h

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#if !defined(INSIDE_ENCLAVE) || defined(VIRTUAL_ENCLAVE)
+#include <array>
+
+namespace ccf
+{
+  static constexpr size_t attestation_report_data_size = 32;
+  using attestation_report_data =
+    std::array<uint8_t, attestation_report_data_size>;
+  static constexpr size_t attestation_measurement_size = 48;
+  using attestation_measurement =
+    std::array<uint8_t, attestation_measurement_size>;
+
+  // From https://developer.amd.com/sev/
+  constexpr auto amd_milan_root_signing_public_key = R"(-----BEGIN PUBLIC KEY-----
+MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA0Ld52RJOdeiJlqK2JdsV
+mD7FktuotWwX1fNgW41XY9Xz1HEhSUmhLz9Cu9DHRlvgJSNxbeYYsnJfvyjx1MfU
+0V5tkKiU1EesNFta1kTA0szNisdYc9isqk7mXT5+KfGRbfc4V/9zRIcE8jlHN61S
+1ju8X93+6dxDUrG2SzxqJ4BhqyYmUDruPXJSX4vUc01P7j98MpqOS95rORdGHeI5
+2Naz5m2B+O+vjsC060d37jY9LFeuOP4Meri8qgfi2S5kKqg/aF6aPtuAZQVR7u3K
+FYXP59XmJgtcog05gmI0T/OitLhuzVvpZcLph0odh/1IPXqx3+MnjD97A7fXpqGd
+/y8KxX7jksTEzAOgbKAeam3lm+3yKIcTYMlsRMXPcjNbIvmsBykD//xSniusuHBk
+gnlENEWx1UcbQQrs+gVDkuVPhsnzIRNgYvM48Y+7LGiJYnrmE8xcrexekBxrva2V
+9TJQqnN3Q53kt5viQi3+gCfmkwC0F0tirIZbLkXPrPwzZ0M9eNxhIySb2npJfgnq
+z55I0u33wh4r0ZNQeTGfw03MBUtyuzGesGkcw+loqMaq1qR4tjGbPYxCvpCq7+Og
+pCCoMNit2uLo9M18fHz10lOMT8nWAUvRZFzteXCm+7PHdYPlmQwUw3LvenJ/ILXo
+QPHfbkH0CyPfhl1jWhJFZasCAwEAAQ==
+-----END PUBLIC KEY-----
+)";
+
+  // Based on the SEV-SNP ABI Spec document at
+  // https://www.amd.com/system/files/TechDocs/56860.pdf
+
+  // Table 3
+  union tcb_version {
+    struct _tcb_version {
+      uint8_t boot_loader;
+      uint8_t tee;
+      uint8_t reserved[4];
+      uint8_t snp;
+      uint8_t microcode;
+    };
+    uint64_t raw;
+  };
```

Comment:
I like anonymous structs, but I think the union here is unnecessary if we're only reading from `.raw`. I think this should be:

```
struct TcbVersion { uint8_t boot_loader; ... }
static_assert(sizeof(TcbVersion) == sizeof(uint64_t), "Can't cast TcbVersion to uint64_t");
```

Then anywhere you want to read this memory as a uint64, you instead do an explicit memory cast (`*(uint64_t*)(&tcb_version)`).

Reply:
@eddyashton the union is about the same to define, and slightly shorter and neater at call sites though? I don't feel strongly about this.

Reply:
If we actually use the union, it's technically undefined behaviour (reading from any union member other than the last written). It's handled sensibly by all known compilers, but still a risk (and might be flagged by smarter tooling in future). With all of these, I think the purpose of the struct should be to expose/decompose the internal fields, and the fact they're efficiently packed/serializable is separate, not intrinsic. If the call site gets too ugly, we can wrap it in a method (maybe all of these structs deserve something like a `std::range<uint8_t, N> raw() const` member?)

Reply:
I also think it's easier to read this way (with union), plus I don't see a downside of using the union here?

Reply:
@eddyashton huh, indeed, it's only allowed in C99, but explicitly UB in C++ since '11. Very unfortunate.

Reply:
If everyone is happy with version I have I will resolve this thread

Reply:
TBH, I think a class with a private `uint8_t bs[8];` and accessor functions like `uint8_t tee() const { return bs[1]; }` would look even neater to me :-)